great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -191,6 +191,62 @@
 T = TypeVar("T", dict, list, str)
 
 
+class DatasourceDict(UserDict):
+    def __init__(self, context: AbstractDataContext):
+        self._context = context
+        self._datasource_store = context._datasource_store
+        self._config_provider = context.config_provider
+        self._fluent_config = context.fluent_config
+
+    @property
+    def data(self):
+        keys = self._datasource_store.list_keys()
+        names = [key.resource_name for key in keys]
+        return {name: self.__getitem__(name) for name in names}
+
+    def __setitem__(self, key: Any, item: Any) -> None:
+        pass
+
+    def __getitem__(self, name: str) -> Any:
+        config = self._datasource_store.retrieve_by_name(name)
+        if "type" in config:
+            return self._init_fluent_datasource(name=name, config=config)
+        return self._init_block_datasource(name=name, config=config)
+
+    def _init_fluent_datasource(self, name: str, config: dict) -> FluentDatasource:
+        if self._datasource_store.cloud_mode:
+            datasource = self._add_fluent_datasource(**config)
+            datasource._rebuild_asset_data_connectors()
+            return datasource
+        return self._fluent_config.get_datasource(name)
+
+    def _init_block_datasource(self, name: str, config: dict) -> Datasource:
+        config = copy.deepcopy(config)  # type: ignore[assignment]
+
+        raw_config_dict = dict(datasourceConfigSchema.dump(config))
+        substituted_config_dict: dict = self._config_provider.substitute_config(
+            raw_config_dict
+        )
+
+        raw_datasource_config = datasourceConfigSchema.load(raw_config_dict)
+        substituted_datasource_config = datasourceConfigSchema.load(
+            substituted_config_dict
+        )
+        substituted_datasource_config.name = name
+
+        return self._context._instantiate_datasource_from_config(
+            raw_config=raw_datasource_config,
+            substituted_config=substituted_datasource_config,
+        )
+
+    def __contains__(self, name: str) -> bool:
+        try:
+            _ = self.__getitem__(name)
+            return True
+        except:
+            return False
```

Comment:
The parent impl of this is `name in self.data` - since `self.data` requires requesting and creating ALL datasources, this impl is more performant.